File: web/tests/reference/binary-trees/construct-binary-tree-from-preorder-and-inorder.js
function buildTree(preorder, inorder) {
  const index = new Map();
  inorder.forEach((v, i) => index.set(v, i));
  let pos = 0;

  const build = (lo, hi) => {
    if (lo > hi) return null;
    const val = preorder[pos++];
    const node = new TreeNode(val);
    const mid = index.get(val);
    node.left = build(lo, mid - 1);
    node.right = build(mid + 1, hi);
    return node;
  };

  return build(0, inorder.length - 1);
}

File: web/tests/reference/binary-trees/construct-binary-tree-from-preorder-and-inorder.py
def build_tree(preorder, inorder):
    index = {v: i for i, v in enumerate(inorder)}
    pos = [0]

    def build(lo, hi):
        if lo > hi:
            return None
        val = preorder[pos[0]]
        pos[0] += 1
        node = TreeNode(val)
        mid = index[val]
        node.left = build(lo, mid - 1)
        node.right = build(mid + 1, hi)
        return node

    return build(0, len(inorder) - 1)
